src/main/java/bridge/domain/bridge/BridgeGame.java

```diff
@@ -0,0 +1,70 @@
+package bridge.domain.bridge;
+
+import bridge.controller.dto.BridgeGameResultDto;
+import bridge.controller.dto.BridgeSquaresDto;
+import java.util.List;
+
+/**
+ * 다리 건너기 게임을 관리하는 클래스
+ */
+public final class BridgeGame {
+
+    private int attemptedCount;
+    private BridgeMap bridgeMap;
+
+    public BridgeGame(final List<String> bridge) {
+        this(1, new BridgeMap(bridge));
+    }
+
+    private BridgeGame(
+            final int attemptedCount,
+            final BridgeMap bridgeMap
+    ) {
+        this.attemptedCount = attemptedCount;
+        this.bridgeMap = bridgeMap;
+    }
+
+    /**
+     * 사용자가 칸을 이동할 때 사용하는 메서드
+     * <p>
+     * 이동을 위해 필요한 메서드의 반환 타입(return type), 인자(parameter)는 자유롭게 추가하거나 변경할 수 있다.
+     */
```

Comment:
`move`메서드를 다리 길이보다 많이 호출하면 `ArrayIndexOutOfBoundsException` 예외가 발생하는 것으로 보입니다.
미션에서는 구현자가 로직의 흐름을 완벽히 제어하고 있기에 문제는 없어보이지만, 협업 관점에서 생각해봤을 때 방어적인 코드를 추가해주는건 어떤지 승호님 생각이 궁금합니다!

가장 간단한 방법은
```suggestion
     public boolean move(final String nextMovePosition) {
         if (isEnded()) {
             return false;
         }
        bridgeMap.recordNextMovePosition(nextMovePosition);
        return true;
    }
```
가 있을 것 같긴한데, 좋은 접근 방법인지는 잘 모르겠네요 ㅎㅎ..

작성하고보니, 구현자에 의도에 맞지 않은 접근일 경우 `(ex. move메서드를 다리 길이보다 많이 호출)`는 그냥 예외를 터트리게 놔두는게 맞나? 라는 생각도 들고, 예외 핸들링 관련은 정말 어려운 주제인 것 같습니다 🤔🤔

Reply:
예외 처리는 제대로 다루기 상당히 어려운 주제라고 생각하고,

현업에 갔을 때 가장 배우고 싶은 주제 중 하나로 꼽고 싶습니다.

외부 입력이 아닌 개발자의 코드 작성에 논리적 오류가 있을 경우에 예외가 발생하는 경우를 좀 더 쉽게 찾기 위해 어떻게 개선할 수 있을지 생각해보면 지금 당장 떠오르는 방법은,

# 1. `assert` statement 사용
```java
public void move(final String nextMovePosition) {
    // isEnded()가 true를 반환할 경우 개발 환경에서 런타임 오류 발생
    assert !isEnded();
    bridgeMap.recordNextMovePosition(nextMovePosition);
}
```

다양한 개발 서적에서 자주 언급되는 `assert` 문이 Java에도 존재하더라구요.

런타임에 발생하는 `OutOfBounds` 등의 논리적 오류를 **오직 개발 및 디버깅 시에만 잡아주고,** 배포 시에는 전부 사라져서 추가적인 성능을 지불할 필요가 없는 것으로 알고 있습니다.

안정적인 프로그램 작성을 위해 `assert` 문을 평균적으로 코드 3줄 당 1번 사용하는 것이 좋다는 이야기를 본 적이 있습니다.

다만 이 `assert`는 예외라는 개념이 존재하지 않거나, 존재해도 실제로 거의 사용되지 않는 `C`, `C++`에서 주로 사용되는 듯 하고, Java에서는 잘 사용되지는 않는 거 같아요.(정확히는 모르겠습니다.)

---

# 2. 논리적 오류 발생 시, 좀 더 명시적인 `Exception` 메세지 작성하기

```java
public void move(final String nextMovePosition) {
    if (isEnded()) {
        throw new IllegalArgumentException("`move` 메소드는 `isEnded` 상태에서 호출 될 수 없습니다.");
    }
    bridgeMap.recordNextMovePosition(nextMovePosition);
}
```

간단한 예외 메세지 추가로 개발 단계에서 논리적인 오류를 잡는데 도움을 줄 수 있을 것 같습니다.

---

# 3. 상태 반환

```java
public MoveResult move(final String nextMovePosition) {
    if (isEnded()) {
        return MoveResult.FAILED;
    }
    bridgeMap.recordNextMovePosition(nextMovePosition);
    return MoveResult.OK;
}
```

`@june-777`님이 제안하신 방법과 거의 비슷하지만 `boolean` 만으로는 정확한 의미를 파악하기 어려울 수 있기 때문에 조금 더 명시적으로 `결과`를 표현하기 위한 enum을 만드는 방법이 있을 것 같습니다.

이것도 역시 `boolean`은 의미 파악의 어려움, `enum`은 추가적인 코드 생성이라는 트레이드 오프가 있는 것 같네요.

---

해당 사례에서 좋은 방법이 무엇인지 생각해봤는데, 논리적인 오류는 개발 단계에서 컨트롤 할 수 있기 때문에 명시적인 메세지와 함께 예외를 터뜨려 주는 것이 사용하는 쪽에 `boolean`이나 `enum` 등을 넘겨주는 것보다 좀 더 깔끔할 것 같다는 생각이 들었습니다.

`상태`를 호출자에 넘긴다는 것은, 예외를 처리할 책임을 호출자에게 위임한다는 것인데, 그러면 `move`를 호출하는 모든 코드에서 반환 받은 `상태`에 따라 처리 로직을 작성해야 하고, 여기 저기 전파되어 유지 보수하는데 어려움이 따를 것 같다는 생각이 들었네요.

실제 현업에서는 예외 처리를 어떤 식으로 하는지 배우기 어렵고, 책에서는 기본적인 개념이나 ~한 방법이 있다 정도로 끝나는 경우가 많은 것 같아서 얼마 전에 공부 겸 다양한 예외 처리 방법에 대해 찾아보고 정리 했었는데, `@june-777`님이 좋은 질문을 던져주신 덕분에 한 번 더 리마인드 할 수 있었네요.😊